### ISSUE TITLE
[Bug]: TA0001 should allow internal procedures

### ISSUE THREAD

**Issue by NKarolak** [ALCops/Analyzers#434]
### Analyzer

TestAutomationCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

TA0001

### Description

My test codeunit contains a subscriber that can fire only temporarily. In order to bind it manually, I need code in a test function similar to the following one (note that `ThisTest` is a codeunit variable pointing to itself:

```
        ThisTest.SetSomeGlobalVariable(SomeValue); // <---- Requires SetSomeGlobalVariable being internal (or global)
        [...]
        BindSubscription(ThisTest);
        [...]
        UnbindSubscription(ThisTest);
```

TA0001 flags procedure `SetSomeGlobalVariable` because it's global. It should not flag it once I set it to `internal`.

**Comment by Arthurvdv** [comment]
Thanks for the detailed report. The constraint you hit is real: `BindSubscription` needs an instance variable, and AL doesn't allow calling `local` procedures through a variable, even one of the codeunit's own type. So the setter must indeed be at least `internal`.

Our hesitation with exempting `internal` from TA0001: internal procedures are callable from every other object in the same app, so the exemption would let test codeunits accumulate app-internal shared logic — the exact thing this rule guards against.

Before we decide, a question: would it be possible to move the manually-bound subscriber and its state into a separate non-test codeunit? Something like:

```al
codeunit 50101 "My Subscriber Helper"
{
    EventSubscriberInstance = Manual;

    var
        ExpectedValue: Code[20];

    internal procedure SetExpectedValue(Value: Code[20])
    begin
        ExpectedValue := Value;
    end;

    [EventSubscriber(...)]
    local procedure OnSomeEvent(var Result: Code[20])
    begin
        Result := ExpectedValue;
    end;
}
```

```al
    [Test]
    procedure TestSomething()
    var
        Helper: Codeunit "My Subscriber Helper";
    begin
        Helper.SetExpectedValue('X');
        BindSubscription(Helper);
        // ... exercise ...
        UnbindSubscription(Helper);
    end;
```

TA0001 doesn't inspect non-test codeunits, so nothing would be flagged, and the helper stays reusable across test codeunits. As far as we can tell, only UI handlers (ConfirmHandler, PageHandler, ...) must live in the test codeunit itself; event subscribers don't have that restriction.

Is there a reason the subscriber needs to be inside the test codeunit in your scenario? If there's a case this pattern can't cover, that would be a strong argument for (a scoped) exemption.

Generated with Claude Fable 5

**Comment by NKarolak** [comment]
Hey Arthur, sorry for the late reply: For unknown reasons, GitHub does not send me any notification emails anymore from most repositories. I will also answer in the other Issues as well.

Regarding your question: Yes, your proposal works. I've just tested it. So it's a valid workaround.

**Comment by Arthurvdv** [comment]
Thanks for the feedback @NKarolak and great to hear that the workaround is working.
